slang-checks/src/main/java/com/sonarsource/slang/checks/DuplicatedFunctionImplementationCheck.java

```diff
@@ -0,0 +1,116 @@
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package com.sonarsource.slang.checks;
+
+import com.sonarsource.slang.api.BlockTree;
+import com.sonarsource.slang.api.FunctionDeclarationTree;
+import com.sonarsource.slang.api.IdentifierTree;
+import com.sonarsource.slang.api.TopLevelTree;
+import com.sonarsource.slang.api.Tree;
+import com.sonarsource.slang.checks.api.CheckContext;
+import com.sonarsource.slang.checks.api.InitContext;
+import com.sonarsource.slang.checks.api.SecondaryLocation;
+import com.sonarsource.slang.checks.api.SlangCheck;
+import com.sonarsource.slang.visitors.TreeContext;
+import com.sonarsource.slang.visitors.TreeVisitor;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.IntStream;
+import org.sonar.check.Rule;
+
+import static com.sonarsource.slang.utils.SyntacticEquivalence.areEquivalent;
+
+@Rule(key = "S4144")
+public class DuplicatedFunctionImplementationCheck implements SlangCheck {
+
+  private static final String MESSAGE = "Update this function so that its implementation is not identical to \"%s\" on line %s.";
+  private static final String MESSAGE_NO_NAME = "Update this function so that its implementation is not identical to the one on line %s.";
+  private static final int MINIMUM_STATEMENTS_COUNT = 2;
+
+  @Override
+  public void initialize(InitContext init) {
+    init.register(TopLevelTree.class, (ctx, tree) -> {
+      Map<Tree, List<FunctionDeclarationTree>> functionsByParents = new HashMap<>();
+      TreeVisitor<TreeContext> functionVisitor = new TreeVisitor<>();
+      functionVisitor.register(FunctionDeclarationTree.class,
+        (functionCtx, functionDeclarationTree) -> functionsByParents
+          .computeIfAbsent(functionCtx.ancestors().peek(), key -> new ArrayList<>())
+          .add(functionDeclarationTree));
+      functionVisitor.scan(new TreeContext(), tree);
+
+      for (Map.Entry<Tree, List<FunctionDeclarationTree>> entry : functionsByParents.entrySet()) {
+        check(ctx, entry.getValue());
+      }
+    });
+  }
+
+  private static void check(CheckContext ctx, List<FunctionDeclarationTree> functionDeclarations) {
+    Set<FunctionDeclarationTree> reportedDuplicates = new HashSet<>();
+    IntStream.range(0, functionDeclarations.size()).forEach(i -> {
+      FunctionDeclarationTree original = functionDeclarations.get(i);
+      functionDeclarations.stream()
+        .skip(i + 1L)
+        .filter(f -> !reportedDuplicates.contains(f))
+        .filter(DuplicatedFunctionImplementationCheck::hasMinimumLineNumber)
+        .filter(f -> areDuplicatedImplementation(original, f))
+        .forEach(duplicate -> {
+          reportDuplicate(ctx, original, duplicate);
+          reportedDuplicates.add(duplicate);
+        });
+    });
+
+  }
+
+  private static boolean hasMinimumLineNumber(FunctionDeclarationTree function) {
```

Comment:
Could be renamed to something like "hasMinimumSize".

Reply:
Done in 2f63a47c9213ef9997ffb53c8cd9ebeec8b373da